File: CodeStudio/next-greater-element-2/main.py

from typing import List
from collections import deque

def nextGreaterElementII(arr: List[int]) -> List[int]:
    
    n = len(arr)
    ans = [-1] * n
    st = deque()

    for i in range(2*n-1, -1, -1) :

        while(len(st) > 0 and arr[i % n] >= st[-1]) :
            st.pop()
        
        if(i < n) :
            ans[i] = st[-1] if len(st) > 0 else -1
        
        st.append(arr[i % n])
    
    return ans

    

File: CodeStudio/next-greater-element-2/main.cpp

#include <stack>

vector<int> nextGreaterElementII(vector<int>& arr) {

    int n = arr.size();
    vector<int> ans(n, -1);
    stack<int> st;

    for(int i = 2*n-1; i >= 0; i--) {
        
        while(!st.empty() && arr[i % n] >= st.top()) {
            st.pop();
        }

        if(i < n && !st.empty()) {
            ans[i] = st.top();
        }

        st.push(arr[i % n]);

    }

    return ans;

}
